### PR TITLE
Fix a bug during documentation rendering where it does not account for api-name in the path when searching for restepec file

### PR DESCRIPTION
**Pull request by junchuanwang**
Our documentationRender would respond to Options calls by checking the restspec json files from class path using a direct resource name as the file name. 

But some cusomters added a prefix as api-name to pegasus plugin's idl options when generating the idl files, therefore the idl file in class path will have that prefix and documentationRender will not find it for option calls.

This change, is to fix this case by searching for all restspec.json files in the classpath, so that the restspec.json which prefixed by api-names are also taken into account.

Tested by local optionss call, e.g.
`curl "localhost:9000/<service-name>/<resource-name>" -X OPTIONS`

### CODE REVIEW

File: restli-server/src/main/java/com/linkedin/restli/internal/server/model/ResourceModelEncoder.java [hunk 1]
```diff
@@ -314,6 +338,53 @@ public ResourceSchema loadOrBuildResourceSchema(final ResourceModel resourceMode
     }
   }
 
+  private InputStream getResourceStreamBySearchingRestSpec(String resourceName, ClassLoader... classLoaders)
+  {
+    final String REST_SPEC_JSON_SUFFIX = "restspec.json";
```
Comment (zhihaodev) [inline restli-server/src/main/java/com/linkedin/restli/internal/server/model/ResourceModelEncoder.java:343]: Move to the constant of the class?

File: restli-server/src/main/java/com/linkedin/restli/internal/server/model/ResourceModelEncoder.java [hunk 2]
```diff
@@ -314,6 +338,53 @@ public ResourceSchema loadOrBuildResourceSchema(final ResourceModel resourceMode
...
+    if (!resourceName.endsWith(REST_SPEC_JSON_SUFFIX))
+    {
+      return null;
+    }
+    if (_restSpecToClassLoaderMap == null)
+    {
+      _restSpecToClassLoaderMap = new HashMap<>();
+      for(ClassLoader classLoader: classLoaders)
+      {
+        URL[] urls = ((URLClassLoader) classLoader).getURLs();
```
Comment (zhihaodev) [inline restli-server/src/main/java/com/linkedin/restli/internal/server/model/ResourceModelEncoder.java:353]: Are all the classLoaders always `URLClassLoader` in this case?

File: restli-server/src/main/java/com/linkedin/restli/internal/server/model/ResourceModelEncoder.java [hunk 3]
```diff
@@ -109,6 +116,10 @@
 
   private final DataCodec codec = new JacksonDataCodec();
 
+  // Used to cache the mapping between restspec file and
+  // the class loader that can get the resource
+  private Map<String, ClassLoader> _restSpecToClassLoaderMap = null;
```
Comment (zhihaodev) [inline restli-server/src/main/java/com/linkedin/restli/internal/server/model/ResourceModelEncoder.java:121]: Would there be any concurrency concerns here? I'm not sure whether the method `loadOrBuildResourceSchema` can be run concurrently or not.

File: restli-server/src/main/java/com/linkedin/restli/internal/server/model/ResourceModelEncoder.java (hunk 2 above)
Comment (junchuanwang) [reply restli-server/src/main/java/com/linkedin/restli/internal/server/model/ResourceModelEncoder.java:353]: Yes. In fact it should be either AppClassLoader or ExtClassLoader which all extends URLClassLoader

File: restli-server/src/main/java/com/linkedin/restli/internal/server/model/ResourceModelEncoder.java (hunk 3 above)
Comment (junchuanwang) [reply restli-server/src/main/java/com/linkedin/restli/internal/server/model/ResourceModelEncoder.java:121]: There will be no concurreny issue as far as I am aware of. Also I did the testing by single steps and I found the execution is through the loop

File: restli-server/src/main/java/com/linkedin/restli/internal/server/model/ResourceModelEncoder.java (hunk 1 above)
Comment (junchuanwang) [reply restli-server/src/main/java/com/linkedin/restli/internal/server/model/ResourceModelEncoder.java]: moved

**Review by zhihaodev — APPROVED** [review]

**Comment by nickibi** [comment]
I am trying to understanding this change. It used to use resource java classes to generate option call data which shows the resource specs. Now we rely on both java classes and respect.json to generate the spec response？

**Comment by junchuanwang** [comment]
> I am trying to understanding this change. It used to use resource java classes to generate option call data which shows the resource specs. Now we rely on both java classes and respect.json to generate the spec response？

we use restspec.json for options calls. And if not found, we will fall back to use java classes. But using java classes is not desired (see customer complains in detail)

So we need to make sure restspec.json are found in this corner case.

**Comment by nickibi** [comment]
> 

Thanks for the explanation.

**Pull request merged**